FIAT/expansions.py

```diff
@@ -260,52 +293,107 @@ def __init__(self, ref_el, scale=None, variant=None):
             elif scale == "l2 piola":
                 scale = 1.0 / ref_el.volume()
         self.scale = scale
+        self.continuity = "C0" if variant == "bubble" else None
+        self.recurrence_order = 2
+        self._dmats_cache = {}
+        self._cell_node_map_cache = {}
 
     def get_num_members(self, n):
-        D = self.ref_el.get_spatial_dimension()
-        return math.comb(n + D, D)
+        return polynomial_dimension(self.ref_el, n, self.continuity)
 
-    def _mapping(self, pts):
-        if isinstance(pts, numpy.ndarray) and len(pts.shape) == 2:
-            return numpy.dot(self.A, pts) + self.b[:, None]
-        else:
-            m1, m2 = self.A.shape
-            return [sum((self.A[i, j] * pts[j] for j in range(m2)), self.b[i])
-                    for i in range(m1)]
+    def get_cell_node_map(self, n):
+        try:
+            return self._cell_node_map_cache[n]
+        except KeyError:
+            cell_node_map = polynomial_cell_node_map(self.ref_el, n, self.continuity)
+            return self._cell_node_map_cache.setdefault(n, cell_node_map)
+
+    def _tabulate_on_cell(self, n, pts, order, cell=0, direction=None):
+        A, b = self.affine_mappings[cell]
+        ref_pts = apply_mapping(A, b, pts)
+        Jinv = A if direction is None else numpy.dot(A, direction)[:, None]
+        sd = self.ref_el.get_spatial_dimension()
+        phi = dubiner_recurrence(sd, n, order, ref_pts, Jinv,
+                                 self.scale, variant=self.variant)
+        if self.continuity == "C0":
+            phi = C0_basis(sd, n, phi)
+        return phi
 
     def _tabulate(self, n, pts, order=0):
-        """A version of tabulate() that also works for a single point.
-        """
-        D = self.ref_el.get_spatial_dimension()
-        return dubiner_recurrence(D, n, order, self._mapping(pts), self.A, self.scale, variant=self.variant)
+        """A version of tabulate() that also works for a single point."""
+        cell_point_map = compute_cell_point_map(self.ref_el, pts)
+        phis = [self._tabulate_on_cell(n, pts[:, ipts], order, cell=k)
+                for k, ipts in enumerate(cell_point_map)]
+
+        if len(phis) == 1:
+            return phis[0]
+
+        sd = self.ref_el.get_spatial_dimension()
+        results = []
+        num_phis = self.get_num_members(n)
+        cell_node_map = self.get_cell_node_map(n)
+        for r in range(order+1):
+            result = numpy.zeros((num_phis,) + (sd,)*r + pts.shape[1:])
+            for ibfs, ipts, phi in zip(cell_node_map, cell_point_map, phis):
+                shape_indices = tuple(range(sd) for _ in range(r))
+                result[numpy.ix_(ibfs, *shape_indices, ipts)] = phi[r]
+            results.append(result)
+        return tuple(results)
+
+    def tabulate_normal_derivative_jump(self, n, ref_pts, facet, order=1):
```

Comment:
rename -- we support 0-order continuity and higher order checks.  Maybe also rework way things are returned (return lower-order jumps as well since we have them?)